analyzer/QUIC.spicy

```diff
@@ -141,16 +141,21 @@ type VariableLengthInteger = unit {
   # https://datatracker.ietf.org/doc/rfc9000/
   # Section 16 and Appendix A
   #
-  first_byte: bytes &size=1 {
-    local uint8_val = uint8($$.to_uint(spicy::ByteOrder::Big));
-    self.bytes_to_parse = 2**((0xC0 & uint8_val) >> 6);
-    # Re-pack without most significant two bits for later use.
-    self.first_byte = pack(0x3f & uint8_val, spicy::ByteOrder::Big);
+  byte0: uint8 {
+    self.bytes_to_parse = 2**((0xC0 & $$) >> 6);
+    self.result = $$ & 0x3F;
   }
-  remaining_bytes: bytes &size=self.bytes_to_parse  - 1;
 
-  on %done {
-    self.result = (self.first_byte + self.remaining_bytes).to_uint(spicy::ByteOrder::Big);
+  byte1: uint8 if ( self.bytes_to_parse > 1 ) {
+    self.result = (self.result << 8) | $$;
+  }
+
+  byte2: uint8 if ( self.bytes_to_parse > 2 ) {
+    self.result = (self.result << 8) | $$;
+  }
+
+  byte3: uint8 if ( self.bytes_to_parse > 3 ) {
+    self.result = (self.result << 8) | $$;
   }
```

Comment:
You could avoid storing these uninteresting values and the repetition here, e.g.,
```suggestion
  : uint8[self.bytes_to_parse - 1] if (self.bytes_to_parse > 0) foreach {
      self.result = (self.result << 8) | $$;
  }
```

Reply:
You could make this even simpler if you made `self.bytes_to_parse` zero-based, e.g., on init `self.bytes_to_parse = 2**((0xC0 & $$) >> 6) - 1` (probably needs a check that the value is not already zero to avoid underflow). You could then parse the remaining bytes with

```ruby
  : uint8[self.bytes_to_parse] foreach {
      self.result = (self.result << 8) | $$;
  }
```

Reply:
Thanks, nice!

I like the explicit `if` of the first proposal and will go with that. Also fits well with the example in the RFC. Think it needs to be `if ( self.bytes_to_parse > 1)` though.

Reply:
Using `if ( self.bytes_to_parse > 1)` would short-circuit even setting up parsing (for nothing), so it definitely wouldn't hurt.